src/software/simulation/physics/physics_ball.cpp

```diff
@@ -108,16 +110,56 @@ bool PhysicsBall::isTouchingOtherObject() const
     return false;
 }
 
-void PhysicsBall::setInFlightForDistance(double in_flight_distance)
+void PhysicsBall::setInFlightForDistance(double in_flight_distance,
+                                         Angle angle_of_departure)
 {
     in_flight_origin          = position();
     in_flight_distance_meters = in_flight_distance;
+    flight_angle_of_departure = angle_of_departure;
 }
 
-bool PhysicsBall::isInFlight()
+double PhysicsBall::calculateDistanceFromGround() const
 {
-    bool ball_currently_in_flight = in_flight_origin.has_value();
-    if (ball_currently_in_flight)
+    double distance_from_ground = 0.0;
+    if (isInFlight())
+    {
+        // We already know the range and angle of departure, so we can re-arrange
+        // to solve for the theoretical initial speed of the ball. This then lets us
+        // calculate the parabolic trajectory.
+        // https://courses.lumenlearning.com/boundless-physics/chapter/projectile-motion/
+        double x = distance(position(), in_flight_origin.value());
+        double initial_speed_numerator =
+            in_flight_distance_meters *
+            ACCELERATION_DUE_TO_GRAVITY_METERS_PER_SECOND_SQUARED;
+        double initial_speed_denominator = (flight_angle_of_departure * 2).sin();
+        double initial_speed =
+            std::sqrt(initial_speed_numerator / initial_speed_denominator);
+        double y_numerator =
+            ACCELERATION_DUE_TO_GRAVITY_METERS_PER_SECOND_SQUARED * x * x;
+        double y_denominator = 2 * initial_speed * initial_speed *
+                               std::pow(flight_angle_of_departure.cos(), 2);
+        double y = std::tan(flight_angle_of_departure.toRadians()) * x -
+                   (y_numerator / y_denominator);
+
+        if (isTouchingOtherObject())
+        {
+            // If the ball is still in flight and touching another object,
+            // it must be on top of another robot/object and so cannot be
+            // lower than that
+            distance_from_ground = std::max(y, ROBOT_MAX_HEIGHT_METERS);
```

Comment:
Can we add comment explaining what happens if a chip lands on top of an obstacle?

Reply:
This is explained in `updateIsInFlight()`

Reply:
ah didn't see that